xds/src/main/java/io/grpc/xds/internal/sds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.sds;
```

Comment:
Should this class be in this package or in `io.grpc.xds`? Right now it is used only for security but I understand it will eventually be used for routing as well. Also in this PR the class is really used from 'io.grpc.xds`. It will be good to get the correct intended package now unless there are compelling reasons

Reply:
This class selects filter config and is a mix for both sds security and routing config. I'm ok to let it be either side and @ejona86  seems to have a strong preference to let it be sds. So this package wins.

Reply:
It will be good to understand @ejona86 's reasons for this package. AFAICS putting this class in `io.grpc.xds` requires few or no visibility changes and also seems to be the right logical place.

Reply:
Where/when did I express my preference? I'd actually expect the filter chain matching to be in `io.grpc.xds`. Actually using the security provider provided by xds should remain in the current ProtocolNegotiator though (and thus probably means staying in sds for the while).

I'm quite concerned about the visibility issues. We need to avoid having lots of internal stuff be publicly visible in the `io.grpc.xds` package.

Reply:
K. I'll summarize our decision:
1. The `FilterChainSelectorRef` containing both security provider and routing config is created in `XdsServerBuilder` and populated in `XdsServerWrapper` both in `io.grpc.xds`.  This seems fine.
1. The `FilterChainSelector` class is doing the filter chain match, the result is security provider and routing config. The ideally should be in `io.grpc.xds` so the `EnvoyServerprotoData` regarding filters might be package private. But due to 3. The selector and selection result data should be public.
1. The FilterChainMatching protocol negotiator and FilterChainMatchingHandler are using the security result in 2. and would stay in sds.

This might solve the visibility issue, drawback is that the filterChainMatchingHandler job jumps between multiple places and would be hard to reason.

Reply:
> The ideally should be in io.grpc.xds so the EnvoyServerprotoData regarding filters might be package private.

It would be okay if we used a separate attribute for the security provider and had the sds code define that attribute. That avoids a circular dependency as `io.grpc.xds.internal.sds` would not need to reference `io.grpc.xds`.

> But due to 3. The selector and selection result data should be public.

And the normal way to handle that is to make the field itself package-private and then have `Internal*` accessors. But for something like EnvoyServerProtoData that doesn't work so well. In that case, `EnvoyServerProtoData` might deserve to be in `io.grpc.xds.internal`. The packages are a bit of a mess right now so it isn't all that clear what to do when, because it hasn't been designed as a whole.

> The FilterChainMatching protocol negotiator and FilterChainMatchingHandler are using the security result in 2. and would stay in sds.

No, I don't think they are consuming it. They are aware of it and they have to pass it around. But it seems improper for them to live in sds. Sort of like how it'd be improper for XdsClient to live in sds just because it handled some of the security configuration.